### ISSUE TITLE
Vue Warn: Expected Array, got Function

### ISSUE THREAD

**Issue by jgaehring** [storybookjs/storybook#12634]
**Describe the bug**
Using a factory function in my component's default prop, as prescribed [in the docs](https://vuejs.org/v2/guide/components-props.html#Prop-Validation) for Objects and Arrays, I get an error:

```console
[Vue warn]: Invalid prop: type check failed for prop "columns". Expected Number, Array, got Function 
```
I don't see this error in my application code, only when running Storybook.

Here's an abbreviated view of my component:

```js
export default {
  name: 'FarmTiles',
  props: {
    columns: {
      type: [Number, Array],
      default: () => [1, 2, 3],
      validator: val => (
        Array.isArray(val)
          ? val.every(v => typeof v === 'number')
          : typeof val === 'number'
      ),
    },
  },
  // ...
}
```

**To Reproduce**

The repository can be checked out at this commit and ran to reproduce: 
https://github.com/jgaehring/farmOS-client/tree/21bd7a223eff59e939aa44ec240f233730a1c02f

Or I'm happy to create a sample repo if that's easier

**Expected behavior**
Vue shouldn't be evaluating the function before validating the type. If I dig through the stack trace, I find this in `vue.esm.js`:

```js
/**
 * Get the default value of a prop.
 */
function getPropDefaultValue (vm, prop, key) {
  // no default, return undefined
  if (!hasOwn(prop, 'default')) {
    return undefined
  }
  var def = prop.default;
  // warn against non-factory defaults for Object & Array
  if (process.env.NODE_ENV !== 'production' && isObject(def)) {
    warn(
      'Invalid default value for prop "' + key + '": ' +
      'Props with type Object/Array must use a factory function ' +
      'to return the default value.',
      vm
    );
  }
  // the raw prop value was also undefined from previous render,
  // return previous default value to avoid unnecessary watcher trigger
  if (vm && vm.$options.propsData &&
    vm.$options.propsData[key] === undefined &&
    vm._props[key] !== undefined
  ) {
    return vm._props[key]
  }
  // call factory function for non-Function types
  // a value is Function if its prototype is function even across different execution context
  return typeof def === 'function' && getType(prop.type) !== 'Function'
    ? def.call(vm)
    : def
}
```
That last ternary should call the factory function, but it's not. Is it possible Storybook is stringifying it or something?

**Screenshots**
This breaks my layout in Storybook (not the right number of columns):

![image](https://user-images.githubusercontent.com/5556545/94751099-9d3c2480-0355-11eb-8e64-feedeff9142a.png)

**Code snippets**
(see snippets above)

**System**
Environment Info:

  System:
    OS: Linux 5.0 Ubuntu 18.04.5 LTS (Bionic Beaver)
    CPU: (8) x64 Intel(R) Core(TM) i7-1065G7 CPU @ 1.30GHz
  Binaries:
    Node: 14.4.0 - ~/.nvm/versions/node/v14.4.0/bin/node
    npm: 6.14.5 - ~/.nvm/versions/node/v14.4.0/bin/npm
  Browsers:
    Chrome: 85.0.4183.121
    Firefox: 81.0
  npmPackages:
    @storybook/addon-actions: ^6.0.22 => 6.0.22 
    @storybook/addon-essentials: ^6.0.22 => 6.0.22 
    @storybook/addon-links: ^6.0.22 => 6.0.22 
    @storybook/vue: ^6.0.22 => 6.0.22 

**Additional context**
A workaround is to simply always provide a default in my stories. This is what I'm doing for the time being, and it's perfectly acceptable to me, but I think this is a really awesome project and want to help improve it. Again, let me know if I can create a sample repo to reproduce, or help troubleshoot. Kudos to all the folks on your team!

**Comment by shilman** [comment]
I suspect this is the relevant code: https://github.com/storybookjs/storybook/blob/next/addons/docs/src/frameworks/vue/extractArgTypes.ts#L21-L25

Why don't you just define the default to be an array?

```
default: [1, 2, 3],
```

**Comment by jgaehring** [comment]
Thanks for the quick response.

> Why don't you just define the default to be an array?

As the official Vue docs say:

```js
    // Object with a default value
    propE: {
      type: Object,
      // Object or array defaults must be returned from
      // a factory function
      default: function () {
        return { message: 'hello' }
      }
    },
```

I'm not totally sure _why_ Vue requires you to use a factory function to return the default of any any Object/Array props (I found some unsatisfactory answers in vuejs/vue#1032), but it certainly yells at you if you don't. I'd like to avoid that.

Perhaps something like this could remedy it:

```js
      try {
        // eslint-disable-next-line no-eval
        defaultValue = eval(defaultValue);
        if (typeof defaultValue === 'function' && sbType.name !== 'function') {
          defaultValue = defaultValue();
        }
        // eslint-disable-next-line no-empty
      } catch {}
```

I can test this out later today if you think it could work.

**Comment by shilman** [comment]
@jgaehring sorry for the slow reply. confusing why vue enforces that .. but it is what it is. your sounds reasonable to me, tho make sure that this is only applied to object/array types?

**Comment by jgaehring** [comment]
That sounds good. I'm about to head out on a week of vacation but will submit a PR when I get back.

**Comment by shilman** [comment]
@jgaehring awesome. have a wonderful vacation!

**Comment by marsimeau** [comment]
I have the same issue when using factory functions to create default values. I use this pattern to create a random id unique to each instance :

```js
id: {
  type: String,
  default: () => `TextField_${randomString()}`
},
```

This warning is printed in my console :

```
[Vue warn]: Invalid prop: type check failed for prop "id". Expected String with value "() => `TextField_${randomString()}`", got Function
```

When a default value is a `Function` it is always considered a factory by Vue **unless** its prop type is also `Function`.

I believe this pattern exists to prevent the default value from being mutated. If an instance mutates its prop whilst using the default value, all other instances that also uses the default value would be affected.

Is there a reason why Vue isn't handling the default value on its own when using Storybook?

**Comment by shilman** [comment]
We need to revisit the `defaultValue` handling. @tmeasday just did a bunch of bugfixing in this area and came to a similar conclusion I think @marsimeau. Tom, WDYT about this one?

**Comment by tmeasday** [comment]
@shilman I wasn't looking at the Vue defaultValue extraction, but I imagine the semantics @marsimeau mentioned:

> When a default value is a Function it is always considered a factory by Vue unless its prop type is also Function.

Should be taken into account by the vue default value "extractor".

Certainly we should definitely drop "try and use the extracted string, unless we can eval it" logic that you linked to above (https://github.com/storybookjs/storybook/blob/next/addons/docs/src/frameworks/vue/extractArgTypes.ts#L21-L25) for the same reasons we did in the React implementation.

**Comment by jgaehring** [comment]
> Certainly we should definitely drop "try and use the extracted string, unless we can eval it" logic that you linked to above

I gathered there were some deeper issues at play here (see https://github.com/storybookjs/storybook/pull/12795#issuecomment-752240389), which led me to believe that wouldn't be enough to resolve the issue. You can probably close that PR, in that case, since it's sort of a dead-end. Or I can close it.

Dunno if I can help further, since this seems to exceed my understanding of how default values are extracted, but if I can help test at some point, let me know.

**Comment by tmeasday** [comment]
@jgaehring perhaps we can fix this together, either in your PR or a new one?

Let me explain what I did in react-land and perhaps it'll transfer over to vue w/ this tweak.

**Previously**:

1. `react-docgen` (or similar) extracted information about the default value **as a string** and placed it in `defaultSummary.detail` etc.
2. We tried to `eval()` that string to convert it to the correct type
3. If that failed, we fell back to using the actual string value as the default value, for *reasons* that mainly involved working around bugs in `react-docgen`.

This approach was a mistake, because

(a) it isn't the end of the world to just not set the default value, the whole point of a "default"is it isn't required to be set -- the main downside is that any control doesn't get pre-populated.

(b) it is really bad to get a default value wrong. As evidenced by this issue (and many others TBH)

**Now**:

1. We actually try and pull the default value off the component instance. This is usually (but not always sadly) possible in React with `Component.defaultProps`. In such cases we have a pretty bulletproof default value.
2. If that fails, and we think `react-docgen` has a sensible string value, we `eval()` it.

Is it possible to read default values off a component at runtime in Vue? If so we can probably just do 1. and not worry about 2. at all. Then we would just need to replicate Vue's special handling of functional default "values".

**Comment by jgaehring** [comment]
> perhaps we can fix this together, either in your PR or a new one?

Happy to, @tmeasday. I'm over deadline on other work right now, so it might be a week or two, is the only thing. So anyone else who wants to pick this up in the meantime, please feel free!

> Is it possible to read default values off a component at runtime in Vue?

I believe so, on the [`vm.$options`](https://vuejs.org/v2/api/#vm-options) property. So presumably something like this: 

```js
const defaultProps = Component.$options.props[propName].default
```

But I'll have to check and see if that actually works. :crossed_fingers:

**Comment by tmeasday** [comment]
Sounds there is a way forward @jgaehring!

> I'm over deadline on other work right now, so it might be a week or two, is the only thing.

I know the feeling! We've got a couple of weeks until 6.2RC I think, but it would probably good to sort this out before then, otherwise I'll likely forget it all! Let's see how we collectively go.

**Comment by F0rsaken** [comment]
@jgaehring @tmeasday  How's the status about this issue? Do you need help? It's really important for me to make this work, as it's makes some of my stories unusable 😞

**Comment by tmeasday** [comment]
@F0rsaken looks like this missed the boat on 6.2 (although we could patch it if someone does a PR).

In 6.3 we plan on dropping this whole default value thing; so if you can wait a bit for those changes (and use an alpha) it'll be fixed there.

**Comment by mikestu** [comment]
Do you expect the 6.3 alpha with this fix soon?  We have half a dozen Vue components this affects, and holds up our 6.2 upgrade for now.  I would imagine that a lot of people out there have Vue props that require default value factory functions, because they are Array or Object types, so 6.2 may be a no-go for many Vue devs.  If the workaround may work for stories, we still can't take it out of the actual components themselves anyway.  I'm glad a fix may be coming soon, so thanks!

**Comment by jgaehring** [comment]
I think I have time this afternoon to take a stab at some of the suggestions in https://github.com/storybookjs/storybook/issues/12634#issuecomment-783932795 and my following comment. I'll update here.

**Comment by jgaehring** [comment]
I gave this a shot today but I could not figure out where/how to get access to the original Vue component. The `component` argument passed to `extractArgTypes` is not the actual component instance, so it does not have the `$options` property as I described above. Perhaps someone with more knowledge of Storybook's internals can shed a light.

I'm not sure if/when I'll have another chance to work on this, might not be before 6.3 and this could all become moot anyways. Perhaps, in the meantime, the workaround I described at the top could help others until 6.3 is ready.

**Comment by tmeasday** [comment]
@mikestu 

> Do you expect the 6.3 alpha with this fix soon?

Not super soon, there is a bit of UI work required to allow us to drop default values.

>  If the workaround may work for stories, we still can't take it out of the actual components themselves anyway.

I'm not sure I quite understand this, sorry!

**Comment by jgaehring** [comment]
> I'm not sure I quite understand this, sorry!

Sorry, yea, I should have explained more. The workaround I referenced is to always provide `args` in your story for the props that have a factory function default. Example here: https://github.com/farmOS/farmOS-client/commit/afc55bd52a2538311901d14158520416f52a0948. This is a bit onerous as workarounds go, but it keeps things working, w/o having to make changes to your actual application code.

@F0rsaken, @mikestu, would something like that work in your situations?

**Comment by mikestu** [comment]
The issue may be around the default factory function for Vue props, when running in Storybook, and not anything to do with the stories themselves.  I see multiple issues logged in this repo, related to Vue prop defaults as a function, where it seems to not use the result of the factory function correctly.  Here's one, where he also says "The Default parameter should be the result of an expression, not the expression.":  https://github.com/storybookjs/storybook/issues/14434

**Comment by jgaehring** [comment]
Yea, I think I found that if you supply the props in all your stories, Storybook does not evaluate the code related to defaults and therefore no error is thrown and nothing breaks. Again, not an ideal fix, just a workaround, and of course that might not work for everyone.

**Comment by tmeasday** [comment]
Just to be clear, you can bypass SB's default props handling by setting `argTypes.defaultValue` directly on the component. Setting an `arg` value works too, though.

You might ask: what is the difference between a default value and an initial `arg` value? The answer is not much really ;)

**Comment by jgaehring** [comment]
Just coming back to close this issue, since I believe it has been resolved by #14900. I've held off on updating Storybook in my own application, b/c I'll probably be upgrading to Vue 3 first. So I can't verify that PR resolves everything covered here, but I'm optimistic that should do it. If not it probably merits a separate issue.

Thanks again @tmeasday for your support! :relaxed:
